channeldb/messagestore.go

```diff
@@ -0,0 +1,227 @@
+package channeldb
+
+import (
+	"bytes"
+	"encoding/binary"
+	"sort"
+
+	"github.com/boltdb/bolt"
+	"github.com/go-errors/errors"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// basePeerBucketKey is the base key for generating peer bucket keys.
+	basePeerBucketKey = []byte("peermessages")
+)
+
+// MessagesStore represents the boltdb storage for messages inside
+// retransmission sub-system. For every peer we have number of buckets which
+// represent the different type of messages, such logic is due to the fact
+// that by receiving ack messages we need to delete all messages of specific
+// types.
+type MessagesStore struct {
+	// id is a unique identificator of peer.
+	id []byte
+	db *DB
+
+	// cache stores the correspondence of message code and bool which
+	// denotes do messages of this type are exist in storage.
+	cache map[lnwire.Code]bool
+}
+
+// NewMessageStore creates new instance of message storage.
+func NewMessageStore(id []byte, db *DB) *MessagesStore {
+	cache := make(map[lnwire.Code]bool)
+
+	// Initially initialize message cache like if we have all types of
+	// messages.
+	for _, code := range lnwire.AllCommands() {
+		cache[code] = true
+	}
+
+	return &MessagesStore{
+		id:    id,
+		db:    db,
+		cache: cache,
+	}
+}
+
+// Add add message to the storage.
+func (s *MessagesStore) Add(msg lnwire.Message) error {
+	if err := s.db.Batch(func(tx *bolt.Tx) error {
+		peerBucketKey := s.getPeerBucketKey()
+		peerBucket, err := tx.CreateBucketIfNotExists(peerBucketKey)
+		if err != nil {
+			return err
+		}
+
+		// We use integer representation of code as bucket key due to
+		// the fact that lately we will need to restore messages. And
+		// such approach allows as not to do additional steps for
+		// determine the message type.
+		code := msg.Command()
+		key := make([]byte, 4)
+		binary.BigEndian.PutUint32(key, uint32(code))
+
+		messageBucket, err := peerBucket.CreateBucketIfNotExists(key)
+		if err != nil {
+			return err
+		}
+
+		var b bytes.Buffer
+		if err := msg.Encode(&b, 0); err != nil {
+			return err
+		}
+
+		// Generate next sequence number to preserver the message order.
+		sequence, err := peerBucket.NextSequence()
```

Comment:
Hmm, what's implemented here currently isn't quire what we discussed offline. As is now, a sorting step is inserted before retrieving all the messages for a peer as the messages are stored in distinct buckets. 

Instead, what I originally described was this: 
   * A single top-level bucket that maps: `index -> code || msg`. Concatenating the message code to the stored data allows the db logic to properly parse the wire message without trial and error or needing an additional index. 
   * Within this top-level bucket, another bucket would be stored which acts as an index into the top-level bucket. This bucket will be used to locate which messages can be deleted from the log in response to a retrieved ACK message. 

   The mapping for items in this bucket would be: `messageCode -> {index_1, index_2, index_3, etc.}`. So when receiving a new message, you check for the existence of the message code in this index bucket, then delete all the indexes from the top-level bucket that are returned. 

   Similarly, when adding a new message to the top-level bucket, another compile-time constant set of mappings needs to be consulted to determine which message ACKs the message being stored. So in addition to storing it in the top-level bucket, you'd also append to the record for the `messageCode` mappings.